src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,525 @@
+    : impl_{std::unique_ptr<ColumnMetaDataBuilderImpl>(
+          new ColumnMetaDataBuilderImpl(props, column, contents))} {}
+
+ColumnMetaDataBuilder::~ColumnMetaDataBuilder() {}
+
+void ColumnMetaDataBuilder::set_file_path(const std::string& path) {
+  impl_->set_file_path(path);
+}
+
+void ColumnMetaDataBuilder::SetEncodingsWithDictionaryFallback() {
+  impl_->SetEncodingsWithDictionaryFallback();
+}
+
+void ColumnMetaDataBuilder::Finish(int64_t num_values, int64_t dictionary_page_offset,
+    int64_t index_page_offset, int64_t data_page_offset, int64_t compressed_size,
+    int64_t uncompressed_size) {
+  impl_->Finish(num_values, dictionary_page_offset, index_page_offset, data_page_offset,
+      compressed_size, uncompressed_size);
+}
+
+void ColumnMetaDataBuilder::SetStatistics(const ColumnStatistics& result) {
+  impl_->SetStatistics(result);
+}
+
+class RowGroupMetaDataBuilder::RowGroupMetaDataBuilderImpl {
+ public:
+  explicit RowGroupMetaDataBuilderImpl(std::shared_ptr<WriterProperties> props,
+      const SchemaDescriptor* schema, uint8_t* contents)
+      : properties_(props), schema_(schema), current_column_(0) {
+    row_group_ = reinterpret_cast<format::RowGroup*>(contents);
+    initialize_columns(schema->num_columns());
+  }
+  ~RowGroupMetaDataBuilderImpl() {}
+
+  ColumnMetaDataBuilder* NextColumnMetaData() {
+    DCHECK(current_column_ < num_columns())
+        << "The schema only has " << num_columns()
+        << " columns, requested metadata for column: " << current_column_;
+    auto column = schema_->Column(current_column_);
+    auto column_builder = ColumnMetaDataBuilder::GetColumnMetaDataBuilder(properties_,
+        column, reinterpret_cast<uint8_t*>(&row_group_->columns[current_column_++]));
+    auto column_builder_ptr = column_builder.get();
+    column_builders_.push_back(std::move(column_builder));
+    return column_builder_ptr;
+  }
+
+  void Finish(int64_t num_rows) {
+    DCHECK(current_column_ == schema_->num_columns())
+        << "Only " << current_column_ - 1 << " out of " << schema_->num_columns()
+        << " columns are initialized";
+    size_t total_byte_size = 0;
+
+    for (int i = 0; i < schema_->num_columns(); i++) {
+      DCHECK(row_group_->columns[i].file_offset > 0) << "Column " << i
+                                                     << " is not complete.";
+      total_byte_size += row_group_->columns[i].meta_data.total_compressed_size;
+    }
+
+    row_group_->__set_total_byte_size(total_byte_size);
+    row_group_->__set_num_rows(num_rows);
+  }
+
+ private:
+  int num_columns() { return row_group_->columns.size(); }
+
+  void initialize_columns(int ncols) {
+    std::vector<format::ColumnChunk> val;
+    format::ColumnChunk col_chunk;
+    format::ColumnMetaData meta_data;
+    col_chunk.__set_meta_data(meta_data);
+    for (int i = 0; i < ncols; i++) {
+      val.push_back(col_chunk);
+    }
+    row_group_->__set_columns(val);
```

Comment:
Modify `row_group_->columns` in place and set the flag manually, to avoid copying. e.g.

```
row_group_->__isset.columns = true;
row_group_->columns.resize(ncols);
```

and you can similarly twiddle the `__isset.meta_data` bit in the column chunks. You could set this in `ColumnMetaDataBuilder::Finish` too